lib/nuntiux/mocker.ex

```diff
@@ -326,71 +331,121 @@ defmodule Nuntiux.Mocker do
     :ok
   end
 
+  # credo:disable-for-lines:51 Credo.Check.Refactor.ABCSize
   @doc false
-  @spec maybe_run_expects(message, expects) :: expects_matched
+  @spec maybe_run_expects(message, expects, opts) :: mocked_stack_with
         when message: term(),
              expects: expects(),
-             expects_matched: expects_matched()
-  defp maybe_run_expects(message, expects) do
-    Enum.reduce_while(
-      expects,
-      @label_nomatch,
-      fn
-        {_expect_id, _expect_fun}, {@label_match, _matched} = result ->
-          {:halt, result}
-
-        {_expect_id, expect_fun}, @label_nomatch ->
-          try do
-            {:cont, {@label_match, expect_fun.(message)}}
-          catch
-            :error, :function_clause ->
-              {:cont, @label_nomatch}
-          end
+             opts: opts(),
+             mocked_stack_with: mocked_stack_with()
+  defp maybe_run_expects(message, expects, opts) do
+    no_match_fun = fn _msg -> @label_nomatch end
+    expects = Enum.reverse([{:_, no_match_fun} | Map.to_list(expects)])
+
+    mocked_stack_with =
+      Enum.reduce_while(
+        expects,
+        mocked_stack_with(false),
+        fn
+          {expect_id, expect_fun}, %{mocked?: false, stack: stack_trace} ->
+            try do
+              expect_fun.(message)
+            else
+              @label_nomatch ->
+                {:halt, mocked_stack_with(false, stack_trace)}
+
+              match ->
+                {:halt, mocked_stack_with(true, stack_trace, match)}
+            catch
+              class, reason ->
+                stack_elem = %{
+                  expect_id: expect_id,
+                  class: class,
+                  reason: reason,
+                  trace: __STACKTRACE__
+                }
+
+                {:cont, mocked_stack_with(false, stack_elem, nil, stack_trace)}
+            end
+        end
+      )
+
+    case mocked_stack_with do
+      %{mocked?: false, stack: stack_trace, with: nil} when opts.raise_on_nomatch? ->
+        ex_args = [message: "No expectation matched", stack: stack_trace]
+        Logger.error(ex_args)
```

Comment:
Unlike nuntius, I couldn't find a better way to have this show up in the logs. As-is, though, it's not intrusive, because it's still a choice (sure, it's on by the default, but still a choice).